src/node_crypto.cc

```diff
@@ -211,6 +211,95 @@ static int NoPasswordCallback(char* buf, int size, int rwflag, void* u) {
+      OneByteString(env->isolate(), fs ? fs : "")).FromJust();
+  obj->Set(env->isolate()->GetCurrentContext(),
+      env->reason_string(),
+      OneByteString(env->isolate(), rs ? rs : "")).FromJust();
+
+  // SSL has no API to recover the error name from the number, so we
+  // transform "this error" to "ERR_SSL_THIS_ERROR", which ends up being
+  // close to the original error name.
+  std::string code(rs);
+
+  for (auto& c : code) {
+    if (c == ' ')
+      c = '_';
+    else
+      c = ::toupper(c);
+  }
+
+  const char* prefix = "ERR_NONE_";
+  switch(ERR_GET_LIB(err)) {
+    case ERR_LIB_SYS:    prefix = "ERR_SYS_"; break;
+    case ERR_LIB_BN:     prefix = "ERR_BN_"; break;
+    case ERR_LIB_RSA:    prefix = "ERR_RSA_"; break;
+    case ERR_LIB_DH:     prefix = "ERR_DH_"; break;
+    case ERR_LIB_EVP:    prefix = "ERR_EVP_"; break;
+    case ERR_LIB_BUF:    prefix = "ERR_BUF_"; break;
+    case ERR_LIB_OBJ:    prefix = "ERR_OBJ_"; break;
+    case ERR_LIB_PEM:    prefix = "ERR_PEM_"; break;
+    case ERR_LIB_DSA:    prefix = "ERR_DSA_"; break;
+    case ERR_LIB_X509:   prefix = "ERR_X509_"; break;
+    case ERR_LIB_ASN1:   prefix = "ERR_ASN1_"; break;
+    case ERR_LIB_CONF:   prefix = "ERR_CONF_"; break;
+    case ERR_LIB_CRYPTO: prefix = "ERR_CRYPTO_"; break;
+    case ERR_LIB_EC:     prefix = "ERR_EC_"; break;
+    case ERR_LIB_SSL:    prefix = "ERR_SSL_"; break;
+    case ERR_LIB_BIO:    prefix = "ERR_BIO_"; break;
+    case ERR_LIB_PKCS7:  prefix = "ERR_PKCS7_"; break;
+    case ERR_LIB_X509V3: prefix = "ERR_X509V3_"; break;
+    case ERR_LIB_PKCS12: prefix = "ERR_PKCS12_"; break;
+    case ERR_LIB_RAND:   prefix = "ERR_RAND_"; break;
+    case ERR_LIB_DSO:    prefix = "ERR_DSO_"; break;
+    case ERR_LIB_ENGINE: prefix = "ERR_ENGINE_"; break;
+    case ERR_LIB_OCSP:   prefix = "ERR_OCSP_"; break;
+    case ERR_LIB_UI:     prefix = "ERR_UI_"; break;
+    case ERR_LIB_COMP:   prefix = "ERR_COMP_"; break;
+    case ERR_LIB_ECDSA:  prefix = "ERR_ECDSA_"; break;
+    case ERR_LIB_ECDH:   prefix = "ERR_ECDH_"; break;
+    case ERR_LIB_OSSL_STORE:  prefix = "ERR_STORE_"; break;
+    case ERR_LIB_FIPS:   prefix = "ERR_FIPS_"; break;
+    case ERR_LIB_CMS:    prefix = "ERR_CMS_"; break;
+    case ERR_LIB_TS:     prefix = "ERR_TS_"; break;
+    case ERR_LIB_HMAC:   prefix = "ERR_HMAC_"; break;
+    case ERR_LIB_CT:     prefix = "ERR_CT_"; break;
+    case ERR_LIB_ASYNC:  prefix = "ERR_ASYNC_"; break;
+    case ERR_LIB_KDF:    prefix = "ERR_KDF_"; break;
+    case ERR_LIB_SM2:    prefix = "ERR_SM2_"; break;
+    case ERR_LIB_USER:   prefix = "ERR_USER_"; break;
+  }
+
+  obj->Set(env->isolate()->GetCurrentContext(),
+           env->code_string(),
+           OneByteString(env->isolate(), (prefix + code).c_str()))
+    .FromJust();
+
+  return obj;
+}
+
+
+Local<Value> DecorateV(Environment* env, Local<Value> value, unsigned long err) {
+  Local<Object> obj = value->ToObject(env->isolate()->GetCurrentContext())
+    .ToLocalChecked();
```

Comment:
Do you think you could propagate errors, rather than always use `.ToLocalChecked()`/`.FromJust()`?

Reply:
Sorry, I don't understand, could you describe what kind of errors do you think can be propogated?

The `value` was just allected with `CryptoErrorVector::ToException()`, and its set of `CHECK(!IsEmpty())`, `FromJust()` and `ToLocalChecked()` calls look to me like they will fail before this code could be reached.

Reply:
@sam-github Yeah, I’d also say that that code should be changed… generally, it’s not easy to tell whether a V8 call may end up in JS land/may end up throwing an exception or not. It might well be that this particular `ToObject()` call cannot fail, but at least the `->Set(...).FromJust()` calls are something I’d like to eventually have real error handling for at some point.

Reply:
@addaleax I'm happy to make any changes you want here, but you'll have to be more specific, because I don't know how to deal with `ToException()` returning something that isn't in fact an Object, and that its about to pass to `ThrowException()`. I simplified the code to remove wrappers and move error checking closer to the source.

Reply:
@sam-github Basically:

- Instead of returning `Local<Object>` from `DecorateObject`, return `MaybeLocal<Object>` and return empty `MaybeLocal<Object>()`s when one of the `Set()` calls fails.
- Instead of returning `Local<Value>` from `DecorateValue`, return `MaybeLocal<Value>` (or `MaybeLocal<Object>`?), and return if `ToObject` fails or `DecorateObject()` returns an empty `MaybeLocal<>`, e.g.
```c++
MaybeLocal<Object> DecorateValue(Environment* env, Local<Value> value,
                                 unsigned long err) {  // NOLINT(runtime/int)
  Local<Object> obj;
  if (!Value->ToObject(env->isolate()->GetCurrentContext()).ToLocal(&obj))
    return MaybeLocal<Object>();
  return DecorateObject(env, obj, err);
}
```  
- Do not explicitly throw an exception using `ThrowException` if the return value of `DecorateValue` is empty. An empty `Maybe<>` or `MaybeLocal<>` indicates that an exception is already pending, so we don’t need to throw one ourselves.